vicon/udp.h

```diff
@@ -25,109 +24,145 @@
     #include <unistd.h>
 #endif
 
+// BoB robotics includes
+#include "../common/pose.h"
+
+using namespace units::math;
+using namespace units::time;
+using namespace units::velocity;
+
+using namespace BoBRobotics::Pose;
+
 namespace BoBRobotics {
 //----------------------------------------------------------------------------
 // Vicon Typedefines
 //----------------------------------------------------------------------------
 namespace Vicon
 {
-typedef double Vector[3];
+template <class T>
+using Vector = T[3];
 
 //----------------------------------------------------------------------------
 // Vicon::ObjectData
 //----------------------------------------------------------------------------
-//! Simplest object data class - just tracks position and translation
+//! Simplest object data class - just tracks position and attitude
+template <class LengthUnit = meter_t, class AngleUnit = radian_t>
 class ObjectData
+  : public HasPosition<LengthUnit>, public HasAttitude<AngleUnit>
 {
+static_assert(units::traits::is_length_unit<LengthUnit>::value,
+              "LengthUnit must be a length type (e.g. meter_t)");
+static_assert(units::traits::is_angle_unit<AngleUnit>::value,
+              "AngleUnit must be an angle type (e.g. radian_t)");
+
 public:
-    ObjectData() : m_FrameNumber{0}, m_Translation{0.0, 0.0, 0.0}, m_Rotation{0.0, 0.0, 0.0}
+    ObjectData()
+      : m_FrameNumber{ 0 }
+      , m_Position{ 0_m, 0_m, 0_m }
+      , m_Attitude{ 0_rad, 0_rad, 0_rad }
     {
     }
 
     //----------------------------------------------------------------------------
     // Public API
     //----------------------------------------------------------------------------
-    void update(uint32_t frameNumber, const Vector &translation, const Vector &rotation)
+    void update(uint32_t frameNumber, LengthUnit x, LengthUnit y, LengthUnit z,
+                AngleUnit yaw, AngleUnit pitch, AngleUnit roll)
     {
         // Cache frame number
         m_FrameNumber = frameNumber;
 
         // Copy vectors into class
-        std::copy(std::begin(translation), std::end(translation), std::begin(m_Translation));
-        std::copy(std::begin(rotation), std::end(rotation), std::begin(m_Rotation));
+        m_Position[0] = x; m_Position[1] = y; m_Position[2] = z;
+        m_Attitude[0] = yaw; m_Attitude[1] = pitch; m_Attitude[2] = roll;
     }
 
     uint32_t getFrameNumber() const { return m_FrameNumber; }
-    const Vector &getTranslation() const{ return m_Translation; }
-    const Vector &getRotation() const{ return m_Rotation; }
+    Vector<LengthUnit> &getPosition() override { return m_Position; }
```

Comment:
Once again, it may not be possible, but couldn't these functions be templated rather than ``ObjectData``? That way the Vicon code can use whatever units the system uses and only convert (or not) when you read stuff out.